lib/events/discard.go

```diff
@@ -64,38 +66,60 @@ func (d *DiscardAuditLog) StreamSessionEvents(ctx context.Context, sessionID ses
 	return c, e
 }
 
-// DiscardStream returns a stream that discards all events
-type DiscardStream struct{}
+// NewDiscardRecorder returns a [SessionRecorderChecker] that discards events.
+func NewDiscardRecorder() *DiscardRecorder {
+	return &DiscardRecorder{
+		done: make(chan struct{}),
+	}
+}
+
+// DiscardRecorder returns a stream that discards all events
+type DiscardRecorder struct {
+	completed atomic.Bool
+	done      chan struct{}
+}
 
 // Write discards data
-func (*DiscardStream) Write(p []byte) (n int, err error) {
+func (d *DiscardRecorder) Write(p []byte) (n int, err error) {
+	if d.completed.Load() {
+		return 0, trace.BadParameter("stream is closed")
+	}
+
 	return len(p), nil
 }
 
 // Status returns a channel that always blocks
-func (*DiscardStream) Status() <-chan apievents.StreamStatus {
+func (*DiscardRecorder) Status() <-chan apievents.StreamStatus {
 	return nil
 }
 
 // Done returns channel closed when streamer is closed
 // should be used to detect sending errors
-func (*DiscardStream) Done() <-chan struct{} {
-	return nil
+func (d *DiscardRecorder) Done() <-chan struct{} {
+	return d.done
 }
 
 // Close flushes non-uploaded flight stream data without marking
 // the stream completed and closes the stream instance
-func (*DiscardStream) Close(ctx context.Context) error {
-	return nil
+func (d *DiscardRecorder) Close(ctx context.Context) error {
+	return d.Complete(ctx)
 }
 
-// Complete does nothing
-func (*DiscardStream) Complete(ctx context.Context) error {
+// Complete marks the stream as closed
+func (d *DiscardRecorder) Complete(ctx context.Context) error {
+	if !d.completed.Load() {
+		close(d.done)
+	}
+	d.completed.Store(true)
```

Comment:
You'll want to use a compare and swap here, otherwise it's possible that the close call happens twice.
```suggestion
	if !d.completed.CompareAndSwap(false, true) {
		close(d.done)
	}
```

Reply:
Thanks, done